peps/pep-0749.rst

```diff
@@ -390,6 +390,138 @@ More specifically:
+but skips it because it is a data descriptor. Next, it looks in the class dictionaries
+of the classes in its MRO, finds ``X.__annotations__``, and returns it.
+In the second example, there are no annotations anywhere in the MRO, so
+``type.__getattribute__`` falls back to returning the metaclass attribute.
+
+Metaclass behavior with PEP 649
+-------------------------------
+
+With :pep:`649` as originally specified, the behavior of class annotations
+in relation to metaclasses becomes even more erratic, because now ``__annotations__``
+is lazily added to the class dictionary even for classes with annotations.
+The new ``__annotate__`` attribute is also lazily created on classes without
+annotations, which causes further misbehaviors when metaclasses are involved.
+
+The cause of these problems is that we set the ``__annotate__`` and ``__annotations__``
+class dictionary entries only under some circumstances, and rely on descriptors
+defined on :py:class:`type` to fill them in if they are not set. This approach
+breaks down in the presence of metaclasses, because entries in the metaclass's own
+class dictionary can render the descriptors invisible. Two approaches for fixing
+this issue are:
+
+* Ensure that the entry is *always* present in the class dictionary, even if it
+  is empty or has not yet been evaluated. This means we do not have to rely on
+  the descriptors defined on :py:class:`type` to fill in the field, and
+  therefore the metaclass's attributes will not interfere.
+* Ensure that the entry is *never* present in the class dictionary, or at least
+  never added by logic in the language core. This means that the descriptors
+  on :py:class:`type` will always be used, without interference from the metaclass.
+
+The first approach is straightforward to implement for ``__annotate__``: it is
+already always set for classes with annotations, and we can set it explicitly
+to ``None`` for classes without annotations.
+
+``__annotations__`` causes more problems, because it must be lazily evaluated.
+Therefore, we cannot simply always put an annotations dictionary in the class
+dictionary. The alternative approach would be to never set ``__dict__["__annotations__"]``
+and use some other storage to store the cached annotations. This behavior
+change would have to apply even to classes defined under
+``from __future__ import annotations``, because otherwise there could be buggy
+behavior if a class is defined without ``from __future__ import annotations``
+but its metaclass does have the future enabled. As :pep:`649` previously noted,
+removing ``__annotations__`` from class dictionaries also has backwards compatibility
+implications: ``cls.__dict__.get("__annotations__")`` is a common idiom to
+retrieve annotations.
+
+Alex Waygood has suggested an approach that avoids these problems. On class
+creation, ``cls.__dict__["__annotations__"]`` is set to a special descriptor.
```

Comment:
Do `type.__annotate__` and `type.__annotations__` still exist in this approach? (I don't believe they're needed, but best to be explicit)

Reply:
I agree that I don't think they'd be needed with this approach

Reply:
I believe they may be needed for non-heap types. I will experiment with this when I implement the idea.